### PR TITLE
Added falloff values for Heal, Attack & Repair.

### PR DESCRIPTION
**Pull request by ViezeVingertjes**
Correct me if i'm wrong, i was not sure which constant to use but RANGED_HEAL_POWER made sense for towers.

### CODE REVIEW

**Comment by TheRealMaxion** [comment]
TOWER_POWER_HEAL: 400

**Comment by TheRealMaxion** [comment]
And in combat you want the overhealing.

**Comment by anisoptera** [comment]
Agree, use the constant and add `// TODO: combat overhealing` :)

**Comment by TheRealMaxion** [comment]
The actual healing power is also different with range

**Comment by ViezeVingertjes** [comment]
So many variables haha... could you point me to the docs where i can find information about the range dropoff?

**Comment by TheRealMaxion** [comment]
TOWER_OPTIMAL_RANGE: 5,
TOWER_FALLOFF_RANGE: 20,
TOWER_FALLOFF: 0.75
 At range 5 or less it has full power, at range 20 or more it has (1 - 0.75) times less power, and it scales linearly in between.

**Comment by ViezeVingertjes** [comment]
Replaced the static values with a more dynamic (future proof) version. Also renamed the pull request to reflect the current feature. Only Heal is used for now, but at almost no extra cost the attack & repair values are added to be used in the future.

File: src/programs/city/defense.js [hunk 1]
```diff
@@ -8,6 +8,19 @@ class CityDefense extends kernel.process {
...
+      global.TOWER_DAMAGE_EFFECT = []
+      global.TOWER_REPAIR_EFFECT = []
+      global.TOWER_HEAL_EFFECT = []
+
+      for (let i = 0; i < 50; i++) {
+        global.TOWER_DAMAGE_EFFECT[i] = this.calculateWithFallOff(TOWER_POWER_ATTACK, i)
+        global.TOWER_REPAIR_EFFECT[i] = this.calculateWithFallOff(TOWER_POWER_REPAIR, i)
+        global.TOWER_HEAL_EFFECT[i] = this.calculateWithFallOff(TOWER_POWER_HEAL, i)
+      }
+    }
```
Comment (tedivm) [inline src/programs/city/defense.js]: I think these should be defined in `constants` so they can be reused throughout the code.

Comment (tedivm) [reply src/programs/city/defense.js]: They should also get placed in `package.json` so the linter won't toss an error.

File: src/programs/city/defense.js [hunk 2]
```diff
@@ -55,27 +68,36 @@ class CityDefense extends kernel.process {
...
-      for (tower of towers) {
+      let damage = healTarget.hitsMax - healTarget.hits
+
+      for (let tower of towers) {
+        if (damage <= 0) {
+          break
+        }
+
+        const distance = tower.pos.getRangeTo(healTarget.pos)
+        damage -= global.TOWER_HEAL_EFFECT[distance]
```
Comment (ButAds) [inline src/programs/city/defense.js]: might break if the target, for whatever reason, is in another room. I'd suggest clamping the range to the max room size

File: src/programs/city/defense.js [hunk 3]
```diff
@@ -131,6 +153,20 @@ class CityDefense extends kernel.process {
     const hazardTypes = [ATTACK, RANGED_ATTACK, HEAL, WORK, CLAIM]
     return _.some(hostile.body, b => _.include(hazardTypes, b.type))
   }
+
+  calculateWithFallOff (optimalValue, distance) {
+    let factor = 0.0
+
+    if (distance < TOWER_OPTIMAL_RANGE) {
```
Comment (ButAds) [inline src/programs/city/defense.js]: Can also be written as:
```javascript
let effect = optimalValue;
if(distance > TOWER_OPTIMAL_RANGE) {
    if(distance> TOWER_FALLOFF_RANGE) {
        distance= TOWER_FALLOFF_RANGE;
    }
    effect -= effect * TOWER_FALLOFF * (distance - TOWER_OPTIMAL_RANGE) / (TOWER_FALLOFF_RANGE - TOWER_OPTIMAL_RANGE);
}
return Math.floor(effect);
```
Related to the calculation code in the engine
https://github.com/screeps/engine/blob/master/src/processor/intents/towers/attack.js#L34

**Pull request merged**